### ISSUE TITLE
PG error when unscoping specific value in where clause with OR clause

### ISSUE THREAD

**Issue by kimyu92** [rails/rails#32707]
Unscope specific value on where clause causing error
`ActiveRecord::StatementInvalid: PG::ProtocolViolation: ERROR:  bind message supplies 2 parameters, but prepared statement "a7" requires 4`

### Steps to reproduce

Sample Model, execution and query
```ruby
# in Country.rb Rails
has_many :singles, -> { where(citizen_countries: { civil_status: :single }) }, through: :citizen_countries, source: :citizen
has_many :couples, -> { where(citizen_countries: { civil_status: :married }) }, through: :citizen_countries, source: :citizen
```

```ruby
country = Country.find(1)
country.singles.or(country.couples).unscope(where: :civil_status)
```

```sql
 SELECT citizens.* FROM citizens INNER JOIN citizen_countries
  ON citizens.id = citizen_countries.citizen_id WHERE (
    citizen_countries.country_id = $1 AND citizen_countries.civil_status = $2 AND
    citizen_countries.deactivated_at IS NULL OR citizen_countries.country_id = $3 AND 
    citizen_countries.civil_status = $4 AND citizen_countries.deactivated_at IS NULL
  )  [["country_id", 1], ["country_id", 1]]
```

### Expected behavior
not blowing up and match num of variables in sql

### Actual behavior
ActiveRecord::StatementInvalid: PG::ProtocolViolation: ERROR:  bind message supplies 2 parameters, but prepared statement "a7" requires 4

Snippet of stacktrace
```
lib/active_record/connection_adapters/postgresql_adapter.rb:608:in `exec_prepared'
lib/active_record/connection_adapters/postgresql_adapter.rb:608:in `block in exec_cache'
lib/active_record/connection_adapters/abstract_adapter.rb:590:in `block in log'
lib/active_record/connection_adapters/abstract_adapter.rb:583:in `log'
lib/active_record/connection_adapters/postgresql_adapter.rb:607:in `exec_cache'
lib/active_record/connection_adapters/postgresql_adapter.rb:591:in `execute_and_clear'
lib/active_record/connection_adapters/postgresql/database_statements.rb:103:in `exec_query'
lib/active_record/connection_adapters/abstract/database_statements.rb:377:in `select_prepared'
lib/active_record/connection_adapters/abstract/database_statements.rb:39:in `select_all'
lib/active_record/connection_adapters/abstract/query_cache.rb:95:in `select_all'
lib/active_record/querying.rb:39:in `find_by_sql'
lib/active_record/relation.rb:706:in `exec_queries'
lib/active_record/association_relation.rb:32:in `exec_queries'
lib/active_record/relation.rb:583:in `load'
lib/active_record/relation.rb:260:in `records'
```

### System configuration
**Rails version**: 5.0.7

**Ruby version**: 2.4.4

**Comment by y-yagi** [comment]
Can you create a reproducible bug script from this [template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb)?

**Comment by kimyu92** [comment]
@y-yagi I have created a gist 👉 [here](https://gist.github.com/kimyu92/0d925c9fa81864469af45d23ffb06a21) based on the template provided above. 

The bug is reproducible with either sqlite or pg adapter.
Please let me know if you need further information

**Comment by y-yagi** [comment]
Thanks.  
But this issue already fixed since Rails 5.1.0. 

```ruby 
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'

  gem 'rails', '5.1.0'
  gem 'sqlite3'
  # gem 'pg', '~> 1.0'
  gem 'pry'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'
require 'pry'


ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')

# before uncomment, createdb test
# ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'test')

ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  drop_table :citizen_countries, if_exists: true
  drop_table :citizens, if_exists: true
  drop_table :countries, if_exists: true

  create_table :countries, force: true do |t|
    t.timestamps null: false
    t.string :name, null: false
  end

  create_table :citizens, force: true do |t|
    t.timestamps null: false
    t.string :first_name, null: false
    t.string :last_name, null: false
    t.string :gender, null: false
  end

  create_table :citizen_countries, force: true do |t|
    t.timestamps null: false
    t.references :country, null: false, foreign_key: true
    t.references :citizen, null: false, foreign_key: true
    t.string :civil_status, null: false, default: 'single'
  end
end

class Country < ActiveRecord::Base
  has_many :citizen_countries, inverse_of: :country, dependent: :destroy
  has_many :citizens, through: :citizen_countries
  has_many :singles, -> { where(citizen_countries: { civil_status: :single }) }, through: :citizen_countries, source: :citizen
  has_many :couples, -> { where(citizen_countries: { civil_status: :married }) }, through: :citizen_countries, source: :citizen
end

class Citizen < ActiveRecord::Base
  enum gender: { male: 'male', female: 'female' }
  has_many :countries, through: :citizen_countries
  has_many :citizen_countries, inverse_of: :citizen, dependent: :destroy
end

class CitizenCountry < ActiveRecord::Base
  enum civil_status: { single: 'single', married: 'married' }
  belongs_to :country, inverse_of: :citizen_countries
  belongs_to :citizen, inverse_of: :citizen_countries
end

class BugTest < Minitest::Test
  def test_association_stuff
    country = Country.create!(name: 'Riverhaven')
    person1 = Citizen.male.create!(first_name: 'Harry', last_name: 'Porter')
    person2 = Citizen.female.create!(first_name: 'Hermione', last_name: 'Granger')
    country.citizens << person1
    country.citizens << person2

    country.singles.or(country.couples).unscope(where: :civil_status) # this line works
    p country.singles.or(country.couples).unscope(where: :civil_status).where(citizen_countries: { civil_status: :single }).first # this line blows up
  end
end 
``` 

``` 
$ ruby 32707.rb
#<Citizen id: 1, created_at: "2018-05-02 22:22:55", updated_at: "2018-05-02 22:22:55", first_name: "Harry", last_name: "Porter", gender: "male"> 
```
 
Please try to upgrade. Thanks.

**Comment by kimyu92** [comment]
Thanks for your suggestion. I'm surprised by how many bug fixes not being back-ported to rails5.
